### ISSUE TITLE
clarify how to name fields and groups

### ISSUE THREAD

**Issue by prjemian** [nexusformat/definitions#544]
At the 2017-03-08 [web conference](http://www.nexusformat.org/Telco_20170308.html), the recommendations and rules for naming fields and groups were discussed.  The manual says one thing in the text and another in the regular expression for the recommended practice.

Specifically, the question is do we recommend (or even allow) the use of an *underscore* character (`_`) as the first character of a field or group name?  The text does not describe this while the regular expression, added later, does.  It is clear the recommendation is to use lower case letters.  

The text does not state explicitly that field or group names must not begin with a number.  This should be stated clearly.

There is a second regular expression with accompanying text, that adds uppercase letters to the list of allowed starting characters in a field or group name.  There were no concerns raised during the meeting about this.

**Comment by prjemian** [comment]
Here's the location in the manual: http://download.nexusformat.org/doc/html/datarules.html#index-2

**Comment by prjemian** [comment]
This can be resolved by the next milestone.  Discussion leans towards changing the regex to match the text that says start with a letter.

**Comment by prjemian** [comment]
Do this with a branch+PR

**Comment by prjemian** [comment]
@vasole:

While it looks as if the [manual's text](http://download.nexusformat.org/doc/html/datarules.html#naming-conventions) is in agreement with the regular expressions, there is still [question](https://github.com/silx-kit/silx/issues/2691#issuecomment-515794049) about whether or not this naming convention is the rule.  The question today arises from this text: *It is recommended that all group and field names contain only these characters*  This text is not clear.

In addition to asserting that names must not begin with a number, the manual must assert that names *must* (as in: are required to match) either of the two regular expressions (which also means no `.` characters in the name).  A match with the first expression is preferred.

**Comment by prjemian** [comment]
The [intent of the naming convention](https://github.com/nexusformat/definitions/issues/92) is that groups and field use names that are *valid program variable names*.

Make this point in the documentation.

**Comment by prjemian** [comment]
2020 Code Camp decided this should be taken up at the next NIAC

**Comment by vasole** [comment]
At the ESRF we have serious difficulties to give up starting a group name by a number.

In my opinion the evolution from recommendation to rule comes really too late considering the former specification has been around about 10 years.

**Comment by prjemian** [comment]
@vasole:  You make two separate points.

1. ESRF is invested in naming groups that start with a number.
2. You assert that documentation of the naming convention was a change in the NeXus specification.

About point 1, I understand that and wonder if the NIAC would agree to change the rule.

Let me address point 2 with a reverse chronology:

* 2014 - https://github.com/nexusformat/definitions/issues/92 Use schema to enforce naming rules: *These need to be valid program variable names*
* 2010-02 - [schema rule added](https://github.com/nexusformat/definitions/commit/ed9f56492b46a343e1925a6d5ad91986a44f0e3a) for `validItemName`
* 2010-02 - [regexp pattern added to documentation](https://github.com/nexusformat/definitions/commit/dc2d7052d2883332e786cac9d57ffe1afb0ba77d)
* 2010-02 - regular expression proposed to clarify naming convention (https://groups.google.com/forum/?hl=en#!topic/mahid/v9fJa3sefVk)
  * this discussion started after the ESRF workshop on HDF5 
* 2010-01 - ESRF workshop on HDF5 (http://ftp.esrf.fr/pub/scisoft/HDF5FILES/HDF5_Workshop_2010Jan/)
* 1994 - NeXus proposal documents (see footnotes of https://manual.nexusformat.org/history.html)
  * [NeXus proposal - Page 4](https://www.nexusformat.org/pdfs/NeXus_Proposal.pdf)
  * There is no mention of a naming rule other than how names are constructed.  (see page image below)

![Clipboard01](https://user-images.githubusercontent.com/2279984/85026862-d651df80-b13e-11ea-847a-02f62f6b909b.jpg)

My understanding is that in 2010, it became obvious that NeXus needed to *document* what many on the NIAC had regarded as long-existing rules for naming fields, groups, and attributes in NeXus data files.  Until this documentation was added, these rules were not documented.  (The documentation was based on discussion with NIAC members at the time.)  Examples, such as in the 1994 proposals, only showed names that started with letters.  Numbers were appended to enumerate multiple items.

In short, this was not ever intended to be a change in the NeXus rules but rather to document the existing standard.  Any assertion that this was a change is a misunderstanding due to the lack of previous documentation.

**Comment by vasole** [comment]
The documentation (as written in the web pages) was stating the line below until last year:

_It is recommended that all group and field names contain only these characters_

**Comment by vasole** [comment]
According to the published documentation available until last year, NeXus was basically attribute based while giving freedom to group names.

**Comment by prjemian** [comment]
No.  NeXus did not give freedom to group names.

Here is a definitive statement ([from 8 years ago](https://github.com/nexusformat/definitions/blame/eb3a22c35659e71405d3837a4ebe7778cfbde06c/manual/source/datarules.rst#L13-L21)):

> Group and field Names used within NeXus follow a naming convention which is made up from the following rules: The names of NeXus *group* and *field* items must only contain a restricted set of characters. This set may be described by this regular expression syntax regular expression: ...

**Comment by prjemian** [comment]
Even today, the [manual](https://manual.nexusformat.org/datarules.html?highlight=group%20and%20field) uses the same terms *must only contain* and *It is recommended* in regard to the naming of groups and fields (notably absent is comment on attribute names).  These must be in contradiction since they have provoked this misunderstanding.

![Clipboard01](https://user-images.githubusercontent.com/2279984/85034814-16699000-b148-11ea-8aa3-c8ad7c7ce473.jpg)

**Comment by vasole** [comment]
So you say we are victims of the web pages not reflecting the documentation (It is **recommended** that ...) We always understood that information as NeXus wanting to avoid non-ASCII characters (something we have respected). 

We are starting group names by numbers to make simpler to the user to follow what is going on **irrespective of used software**  particularly when including multiple data-processing steps in the same file.

**Comment by prjemian** [comment]
> web pages not reflecting the documentation 

?

**Comment by vasole** [comment]
To have the quote (8 years ago):

_Group and field Names used within NeXus follow a naming convention which is made up from the following rules: The names of NeXus group and field items must only contain a restricted set of characters. This set may be described by this regular expression syntax regular expression: ..._

While at the same time (still?) having

 _it is recommended to use lower case letters, digits and underscores_  

and

 a _regular expression used to check this **recommendation**_ 

It is what I call not the web pages reflecting the documentation.

**Comment by rayosborn** [comment]
I have documents going back to the mid-nineties when we were first trying to establish some rules (or guidelines) long before NeXus became a formalized standard. This was what we thought sufficient back then:
- Lower case letters are used throughout, except for common symbols and abbreviations such as FWHM. 
- Names are constructed from full words separated by the underscore character e.g. time_of_flight.
- For sequentially indexed group names, the sequential number is simply appended to the name, e.g. filter1, filter2. This convention should be used only for data group names.
- The hierarchical structure of NeXus files should be used to simplify data names, e.g., “temperature” , not “sample_temperature”.

I don't think it occurred to us that anyone would start a name with a digit, because there was an implicit assumption that these names would be mapped on to variable names in a program (Fortran or C). It was before dictionaries were in common use. I think the original version of the Python API used Python attributes to create the hierarchical structure, so you couldn't start any name with a number.

Now that ship has sailed. The Python API allows both attribute and dictionary references, so we can cope with the ESRF convention, and we obviously can't ask them to change already archived files. To be practical, I believe that we should still discourage starting names with numbers, without labelling their use as a standard violation.

**Comment by prjemian** [comment]
Looking at the documentation in question, there are two regular expressions.  It is recommended that rule expressed (*must only contain*) be decided by the first regular expression match.  Otherwise, the second regular expression must match.

What is proposed would change the second regexp to `[\w_]*`.  Here is a [test](https://regex101.com/):

![Clipboard01](https://user-images.githubusercontent.com/2279984/85038455-369b4e00-b14c-11ea-9805-e299b84519af.jpg)
